server/MembraneDetection/MembraneDetection.py

```diff
@@ -0,0 +1,143 @@
+import histomicstk.preprocessing.color_deconvolution as htk_cdeconv
+import histomicstk.filters.shape as htk_shape_filters
+
+import numpy as np
+import scipy as sp
+
+import skimage.color
+import skimage.filters
+import skimage.io
+import skimage.measure
+import skimage.morphology
+
+from ctk_cli import CLIArgumentParser
+
+import logging
+logging.basicConfig()
+
+
+stain_color_map = {
+    'hematoxylin': [0.65, 0.70, 0.29],
+    'eosin':       [0.07, 0.99, 0.11],
+    'dab':         [0.27, 0.57, 0.78],
+    'null':        [0.0, 0.0, 0.0]
+}
+
+
+def main(args):
+
+    #
+    # Read Input Image
+    #
+    print('>> Reading input image')
+
+    im_input = skimage.io.imread(args.inputImageFile)[:, :, :3]
+    size_min = min(im_input.shape[0], im_input.shape[1])
+    im_input = im_input[0:size_min, 0:size_min, :]
+
+    #
+    # Perform color deconvolution -unnormalized
+    #
+    print('>> Performing color deconvolution')
+
+    stain_color_1 = stain_color_map[args.stain_1]
+    stain_color_2 = stain_color_map[args.stain_2]
+    stain_color_3 = stain_color_map[args.stain_3]
+
+    w = np.array([stain_color_1, stain_color_2, stain_color_3]).T
+
+    im_stains = htk_cdeconv.color_deconvolution(im_input, w).Stains
+
+    im_membrane_stain = im_stains[:, :, 2].astype(np.float)
+
+    #
+    # Perform membrane enhancement
+    #
+    print('>> Performing membrane enhancement')
+
+    # membrane filtering
+    im_deviation, im_thetas = htk_shape_filters.membranefilter(
+        im_membrane_stain, [1, 2, 3], 4, 2
+    )
+    im_deviation = 255 * im_deviation / im_deviation.max()
+
+    # segmentation
+    im_mask = im_deviation > args.foreground_threshold * \
+        skimage.filters.threshold_otsu(im_deviation)
+    im_opened = sp.ndimage.binary_opening(
+        im_mask, structure=np.ones((3, 3))
+    ).astype(np.int)
+
+    # skeletonization
+    im_skeleton = skimage.morphology.skeletonize(im_opened)
+
+    #
+    # Perform membrane label detection
+    #
+    print('>> Performing membrane label detection')
+
+    # set default branch mask
+    branch_mask = np.ones((3, 3))
+
+    # perform convolution
+    im_branch_mask = sp.signal.convolve2d(
+        im_skeleton, branch_mask, boundary='wrap', mode='same'
+    )
+
+    im_branches = np.zeros_like(im_branch_mask)
+    im_branches[np.where(im_branch_mask > args.branch_threshold)] = 1
+
+    # label detection
+    im_split = im_skeleton & ~sp.ndimage.binary_dilation(
+        im_branches,
+        structure=skimage.morphology.disk(args.branch_split_dilation)
+    )
+    im_split = sp.ndimage.binary_dilation(
+        im_split,
+        structure=skimage.morphology.disk(args.branch_merge_dilation)
+    )
```

Comment:
@cdeepakroy just wanted to make sure that the code the right above is not to make an erosion but to dilate only,